bridges/snowbridge/primitives/router/src/inbound/v2.rs

```diff
@@ -0,0 +1,642 @@
+		let eth: XcmAsset =
+			(fee_asset.clone(), message.execution_fee.saturating_add(message.value)).into();
+		let mut instructions = vec![
+			DescendOrigin(PalletInstance(InboundQueuePalletInstance::get()).into()),
+			UniversalOrigin(GlobalConsensus(network)),
+			ReserveAssetDeposited(eth.into()),
+			PayFees { asset: fee },
+		];
+		let mut reserve_assets = vec![];
+		let mut withdraw_assets = vec![];
+
+		let mut refund_surplus_to = origin_account_location;
+
+		if let Some(claimer) = message.claimer {
+			// If the claimer can be decoded, add it to the message. If the claimer decoding fails,
+			// do not add it to the message, because it will cause the xcm to fail.
+			if let Ok(claimer) = Junction::decode(&mut claimer.as_ref()) {
+				let claimer_location: Location = Location::new(0, [claimer.into()]);
+				refund_surplus_to = claimer_location.clone();
+				instructions.push(SetHints {
+					hints: vec![AssetClaimer { location: claimer_location }].try_into().unwrap(),
+				}); // TODO
+			}
+		}
+
+		for asset in &message.assets {
+			match asset {
+				Asset::NativeTokenERC20 { token_id, value } => {
+					let token_location: Location = Location::new(
+						2,
+						[
+							GlobalConsensus(EthereumNetwork::get()),
+							AccountKey20 { network: None, key: (*token_id).into() },
+						],
+					);
+					let asset: XcmAsset = (token_location, *value).into();
+					reserve_assets.push(asset);
+				},
+				Asset::ForeignTokenERC20 { token_id, value } => {
+					let asset_loc = ConvertAssetId::convert(&token_id)
+						.ok_or(ConvertMessageError::InvalidAsset)?;
+					let mut reanchored_asset_loc = asset_loc.clone();
+					reanchored_asset_loc
+						.reanchor(&GlobalAssetHubLocation::get(), &EthereumUniversalLocation::get())
+						.map_err(|_| ConvertMessageError::CannotReanchor)?;
+					let asset: XcmAsset = (reanchored_asset_loc, *value).into();
+					withdraw_assets.push(asset);
+				},
+			}
+		}
+
+		if reserve_assets.len() > 0 {
+			instructions.push(ReserveAssetDeposited(reserve_assets.into()));
+		}
+		if withdraw_assets.len() > 0 {
+			instructions.push(WithdrawAsset(withdraw_assets.into()));
+		}
+
+		// If the message origin is not the gateway proxy contract, set the origin to
+		// the original sender on Ethereum. Important to be before the arbitrary XCM that is
+		// appended to the message on the next line.
+		if message.origin != GatewayProxyAddress::get() {
+			instructions.push(DescendOrigin(
+				AccountKey20 { key: message.origin.into(), network: None }.into(),
+			));
+		}
+
+		// Add the XCM sent in the message to the end of the xcm instruction
+		instructions.extend(message_xcm.0);
+
+		let appendix = vec![
+			RefundSurplus,
+			// Refund excess fees to the claimer, if present, otherwise to the relayer.
+			DepositAsset {
+				assets: Wild(AllOf { id: AssetId(fee_asset.into()), fun: WildFungible }),
+				beneficiary: refund_surplus_to,
+			},
+		];
+
+		instructions.extend(appendix);
```

Comment:
using `instructions.push()` for each of the appendix elements would avoid allocating another intermediary `appendix` vec